### ISSUE TITLE
Edit debugger line with editor

### ISSUE THREAD

**Issue by petobens** [pdbpp/pdbpp#362]
`pdb_class.do_edit` allows us to open the current code line and edit it. I was wondering if you could an option to open the debugger prompt line (i.e the current input) in a similar fashion to bash's `edit-and-execute-command`. My use case is that sometimes I want to evaluate a long expression in the debugger prompt and it would be convenient to edit such expression in vim. Thanks in advance!

**Comment by blueyed** [comment]
Good idea, but not trivial: it's rather in control of the input handler (e.g. pyrepl when using it).

It's possible with ipdb in general (which wraps pdbpp), but requires some manual setup (or this PR (https://github.com/ipython/ipython/pull/12132)).

I'm planning to improve the ipdb/IPython integration soon, which would include that.

btw: I wanted to fix `edit` anyway also to handle editing a file given as an argument (that is not a code object - you know that you can "edit" objects, e.g. functions also?)

**Comment by petobens** [comment]
So basically you use pdbpp within ipdb? How do you set that up? (and from your perspective that's the best debugger experience?)

> you know that you can "edit" objects, e.g. functions also?)

Nice! Didn't know about this

**Comment by blueyed** [comment]
@petobens 
> So basically you use pdbpp within ipdb?

Slowly shifting there I'd say.

> How do you set that up?

The easiest is to install https://github.com/gotcha/ipdb and then use `import ipdb; ipdb.set_trace()`.
Since pdb++ hijacks pdb, ipdb/IPython then also uses it.
I've just merged #349 that will add the "++" to the prompt to indicate that.

Some functions/commands are not going through to pdb++, but e.g. sticky mode also works there then.

> (and from your perspective that's the best debugger experience?)

It's not much better, than pure pdbpp (with pyrepl), but I think in the long run it will be great to have completions from Jedi etc also in the debugger.

I've thought about using python-prompt-toolkit directly in pdbpp, but IPython does a lot of the boilerplate to get Jedi integrated, so it seems to make sense to use it.

**Comment by petobens** [comment]
Thanks for the detailed explanation.

**Comment by goodboy** [comment]
> I've thought about using python-prompt-toolkit directly in pdbpp,

@blueyed I am highly interested in this in order to get tab completion and coloring working for `pdbpp` entered from subprocesses which **are not** attached to a tty/pty.

Would it be ok if I created a new issue to discuss this?

**Comment by blueyed** [comment]
> > I've thought about using python-prompt-toolkit directly in pdbpp,
> 
> @blueyed I am highly interested in this in order to get tab completion and coloring working for `pdbpp` entered from subprocesses which **are not** attached to a tty/pty.
> 
> Would it be ok if I created a new issue to discuss this?

Sure, go ahead if you feel like it.
But as for debugging subprocesses/threads I think it is still more in control / an issue of pdb(pp) itself to handle multiple inputs (i.e. use locks).  I have some local WIP for this (more about threads than subprocesses though).

**Comment by goodboy** [comment]
> Sure, go ahead if you feel like it.

@blueyed sweet.

For further details on the general progress on this there's a bunch of linked issues scattered about here and #351.

> But as for debugging subprocesses/threads I think it is still more in control / an issue of pdb(pp) itself to handle multiple inputs (i.e. use locks). I have some local WIP for this (more about threads than subprocesses though).

Over at `tractor` we've actually got a mult-process debugger system working "natively" with `pdbpp` now (and it's quite nice) using exactly this technique: a lock around the root process's tty. 

If interested, [feel free to take a peek](https://github.com/goodboy/tractor/blob/master/tractor/_debug.py).

The future plan is to have this same slickness working across hosts - which is likely to be a much bigger feat 😼

**Comment by goodboy** [comment]
>  I have some local WIP for this (more about threads than subprocesses though).

@blueyed btw would def be interested in seeing this work.

**Comment by blueyed** [comment]
@goodboy it's deep down in my stashes by now - I've done it when working with thread/asyncio-based projects quite a while ago.  Therefore it's not really easy to push it somewhere etc.

**Comment by goodboy** [comment]
gotcha, no worries.

Like I said, already have a working solution but more code to reference is always good!
